File: csharp/dnd-character/DndCharacter.cs
using System;
using System.Linq;

public class DndCharacter
{
    public int Strength { get; }
    public int Dexterity { get; }
    public int Constitution { get; }
    public int Intelligence { get; }
    public int Wisdom { get; }
    public int Charisma { get; }
    public int Hitpoints { get; }

    static readonly Random _rnd = new Random();

    public DndCharacter()
    {
        Strength = Ability();
        Dexterity = Ability();
        Constitution = Ability();
        Intelligence = Ability();
        Wisdom = Ability();
        Charisma = Ability();
        Hitpoints = 10 + Modifier(Constitution);
    }

    public static int Modifier(int score) => (int)Math.Floor((score - 10) / 2.0);

    public static int Ability() =>
        Enumerable
            .Range(0, 4)
            .Select(_ => _rnd.Next(1, 7))
            .OrderBy(v => v)
            .Skip(1)
            .Sum();

    public static DndCharacter Generate() => new DndCharacter();
}

File: fsharp/dnd-character/DndCharacter.fs
module DndCharacter

open System

let modifier x =
    (float (x - 10)) / 2.0
    |> Math.Floor
    |> int

let ability() =
    let rnd = Random()

    [ 1..4 ]
    |> List.map (fun _ -> rnd.Next(1, 7))
    |> List.sort
    |> List.skip 1
    |> List.sum

type DndCharacter() =
    member val Strength = ability()
    member val Dexterity = ability()
    member val Constitution = ability()
    member val Intelligence = ability()
    member val Wisdom = ability()
    member val Charisma = ability()
    member this.Hitpoints with get() = 10 + modifier(this.Constitution)